### DISCUSSION TITLE
Extend other than BaseController?

### DISCUSSION THREAD

**Discussion by gphg** [codeigniter4/shield#656] [Q&A] [answered]
What if my application's base abstraction controller is not `BaseController`, but something else? The Shield's controllers won't able to extended it.

My way to fix it is by rewrite the package's controller into application's.

**Comment by kenjis** [comment]
> My way to fix it is by rewrite the package's controller into application's.

No problem. You can fully customize the controller.

**Reply by iamsyh** [reply]
how do we go about doing this?

**Reply by datamweb** [reply]
https://codeigniter4.github.io/shield/customization/#extending-the-controllers

**Reply by iamsyh** [reply]
i am already doing that but i get the following error:

```
Error

Class "App\Controllers\BaseController" not found

/home/my_dir/vendor/codeigniter4/shield/src/Controllers/LoginController.php at line 13
```

**Reply by datamweb** [reply]
Can you explain the steps you did step by step?
Write the codes so that I copy and paste.

**Reply by iamsyh** [reply]
Okay so I changed the default namespace for my codeigniter app here:

Constants.php:
```php
/*
 | --------------------------------------------------------------------
 | App Namespace
 | --------------------------------------------------------------------
 |
 | This defines the default Namespace that is used throughout
 | CodeIgniter to refer to the Application directory. Change
 | this constant to change the namespace that all application
 | classes should use.
 |
 | NOTE: changing this will require manually modifying the
 | existing namespaces of App\* namespaced-classes.
 */
defined('APP_NAMESPACE') || define('APP_NAMESPACE', 'MyAppNamespace\Backend\App');
```

So now my BaseController has the namespace: `MyAppNamespace\Backend\App\Controllers;`

Now when I try to extend ShieldLogin in my custom LoginController.php, it doesn't work:

```php
<?php
namespace MyAppNamespace\Backend\App\Controllers\Common;

use CodeIgniter\Shield\Controllers\LoginController as ShieldLogin;

class LoginController extends ShieldLogin
{
    protected $helpers = ['settings', 'misc'];

    public function getValidationRules(): array
    {
        return setting('Validation.login') ?? [
            'username' => [
                'label' => 'Auth.username',
                'rules' => config('AuthSession')->usernameValidationRules,
            ],
            'password' => [
                'label' => 'Auth.password',
                'rules' => 'required',
            ],
        ];
    }
}
```

throws:

```
Error

Class "App\Controllers\BaseController" not found

/home/my_dir/vendor/codeigniter4/shield/src/Controllers/LoginController.php at line 13
```

**Reply by datamweb** [reply]
The error received seems to be because the shield controllers inherit from `BaseController`.

https://github.com/codeigniter4/shield/blob/9d1f6edb51e57dc9f235c00cca9869796400a0ad/src/Controllers/ActionController.php#L7
https://github.com/codeigniter4/shield/blob/9d1f6edb51e57dc9f235c00cca9869796400a0ad/src/Controllers/LoginController.php#L7
https://github.com/codeigniter4/shield/blob/9d1f6edb51e57dc9f235c00cca9869796400a0ad/src/Controllers/MagicLinkController.php#L7
https://github.com/codeigniter4/shield/blob/9d1f6edb51e57dc9f235c00cca9869796400a0ad/src/Controllers/RegisterController.php#L7

When it changes `APP_NAMESPACE` the system cannot find the file `BaseController` it uses.

Solution, it seems that there is no problem by copying a copy of `BaseController` and set new  namespace and setting new  namespace  it in the shield controllers(**ActionController**, **LoginController**, **MagicLinkController**, **RegisterController**). 
**However, this method is not useful and does not seem like a principled solution. So I don't know, sorry!**
Maybe someone else can help. :

**Reply by iamsyh** [reply]
yes I was thinking the same, but would not want to create a copy of LoginController from shield nor modify this file directly..

Thank you for looking into it though! :)

**Reply by kenjis** [reply]
It seems wrong that the Shield controllers extend `App\Controllers\BaseController`.
That is Shield depends on App. But it is the opposite direction of dependence.

**Reply by lonnieezell** [reply]
What's the other option? To not have any dependency? In that case you would lose out on any app-level setup people might do in their BaseController. I am thinking being able to rename the APP namespace might be a good thing to jettison for v5, though.

**Reply by kenjis** [reply]
I don't know this is good or bad, how about moving the all logic in a Shield controller to a trait?

**Reply by lonnieezell** [reply]
Not a fan of that option. 

I don't know there is a great solution for this, honestly. In CodeIgniter's user guide we recommend extending BaseController for all controllers, so I still think it makes sense to handle that here. 

As much as I hate to say it, I think the solution may be to add a note to the docs for this specific situation, saying that you'll have to copy the controllers and extend the proper base controller.

**Reply by gphg** [reply]
I thought things such as class alias or psr mapping (done in composer.json) can achieved this.

**Reply by lonnieezell** [reply]
I hadn't really considered that, but ... maybe? 

`class_alias` will always load that file on every request whether needed or not, IIRC. 

PSR mapping might work - but I would have to try it. I think you would still end up with conflicts since the files have the `App` namespace hard-coded.

**Reply by kenjis** [reply]
What is PSR mapping?

If you can use `App\Controllers\BaseController` as an alias, it seems better to use `App` as your app namespace.

**Comment by kenjis** [comment, marked as answer]
There is no need to inherit the controllers of Shield.
It is possible to write your controllers from scratch, or just copy the controllers of Shield and modify them as you like.

After all, the controllers in Shield are sample code, and the controllers in your app are essentially yours.
